server/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -147,4 +159,35 @@ public final String simpleName() {
     /** Return the merge of {@code mergeWith} into this.
      *  Both {@code this} and {@code mergeWith} will be left unmodified. */
     public abstract Mapper merge(Mapper mergeWith);
+
+    static class ColumnPositionResolver {
+        private final Map<Integer, Mapper> unpositionedMappers = new TreeMap<>(Comparator.naturalOrder());
+        private int maxColumnPosition = 0;
+
+        public ColumnPositionResolver resolve(@Nonnull ColumnPositionResolver toResolve) {
+            // only toResolve needs to hold a list of mappers to resolve the col positions
+            assert this.unpositionedMappers.size() == 0;
+            int maxColumnPosition = Math.max(this.maxColumnPosition, toResolve.maxColumnPosition);
+            for (var e : toResolve.unpositionedMappers.values()) {
+                e.position = ++maxColumnPosition;
+            }
+
+            var merged = new ColumnPositionResolver();
+            merged.maxColumnPosition = maxColumnPosition;
+            return merged;
+        }
+
+        public void addUnpositionedMapper(BuilderContext context, Mapper mapper) {
+            this.unpositionedMappers.put(
+                // mappers are input in depth first order but want to convert it to
+                // breadth-first, then insertion order second. This way, parent's position < children's positions.
+                // If the number of unpositioned mappers are > 1000, this may break.
+                context.contentPath.currentDepth() * 1000 + this.unpositionedMappers.size(),
```

Comment:
Where is this limit coming from?

Reply:
It is limitation of my implementation...
Came up with a new one,
`(long) context.contentPath.currentDepth() * Integer.MAX_VALUE + this.unpositionedMappers.size()` so that each key can be unique.